balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,626 @@
+		b.subConns.Set(curAddr, sd)
+	}
+
+	scd := sd.(*scData)
+	switch scd.state {
+	case connectivity.Idle:
+		scd.subConn.Connect()
+	case connectivity.TransientFailure:
+		if !b.addressIndex.increment() {
+			b.endFirstPass(scd.lastErr)
+			return
+		}
+		b.requestConnection()
+	case connectivity.Ready:
+		// Should never happen.
+		b.logger.Errorf("Requesting a connection even though we have a READY subconn")
+	}
+}
+
+// updateSubConnState handles subConn state updates.
+// Only executed in the context of a serializer callback.
+func (b *pickfirstBalancer) updateSubConnState(sd *scData, state balancer.SubConnState) {
+	// Previously relevant subconns can still callback with state updates.
+	// To prevent pickers from returning these obsolete subconns, this logic
+	// is included to check if the current list of active subconns includes this
+	// subconn.
+	if activeSD, found := b.subConns.Get(sd.addr); !found || activeSD != sd {
+		return
+	}
+	if state.ConnectivityState == connectivity.Shutdown {
+		return
+	}
+
+	if state.ConnectivityState == connectivity.Ready {
+		b.shutdownRemaining(sd)
+		b.addressIndex.seekTo(sd.addr)
+		b.state = connectivity.Ready
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+		return
+	}
+
+	// If we are transitioning from READY to IDLE, reset index and re-connect when
+	// prompted.
+	if b.state == connectivity.Ready && state.ConnectivityState == connectivity.Idle {
+		// Once a transport fails, we enter idle and start from the first address
+		// when the picker is used.
+		b.state = connectivity.Idle
+		b.addressIndex.reset()
+		b.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.Idle,
+			Picker:            &idlePicker{exitIdle: b.ExitIdle},
+		})
+		return
+	}
+
+	if b.firstPass {
+		switch state.ConnectivityState {
+		case connectivity.Connecting:
+			// We can be in either IDLE, CONNECTING or TRANSIENT_FAILURE.
+			// If we're in TRANSIENT_FAILURE, we stay in TRANSIENT_FAILURE until
+			// we're READY. See A62.
+			// If we're already in CONNECTING, no update is needed.
+			if b.state == connectivity.Idle {
+				b.cc.UpdateState(balancer.State{
+					ConnectivityState: connectivity.Connecting,
+					Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+				})
+			}
+		case connectivity.TransientFailure:
+			sd.lastErr = state.ConnectionError
+			// Since we're re-using common subconns while handling resolver updates,
+			// we could receive an out of turn TRANSIENT_FAILURE from a pass
+			// over the previous address list. We ignore such updates.
+			curAddr := b.addressIndex.currentAddress()
+			if activeSD, found := b.subConns.Get(curAddr); !found || activeSD != sd {
+				return
+			}
+			if b.addressIndex.increment() {
+				b.requestConnection()
+				return
+			}
+			// End of the first pass.
+			b.endFirstPass(state.ConnectionError)
+		case connectivity.Idle:
```

Comment:
We handled the case where the LB policy is `Ready` and it receives a subchannel update with the ready subconn moving to `Idle`. This is handled in line 444: https://github.com/grpc/grpc-go/pull/7498/files#diff-7bcf0c14316d983066a08e7e895a07871408030144da3f7ce51204f0a5d213c5R444

But we could also get here while we are still in the first pass, and let's say we started connecting to addresses `{1, 2, 3 ... N}`, and none of them have become `Ready` yet. And now, the subconn for `addr-1` moves from `TF` to `Idle` after its backoff expires. Don't we have to handle that case here, by doing nothing (instead of resetting the address index and returning an idle picker)?